src/Microsoft.OData.Core/ODataInnerError.cs

```diff
@@ -33,38 +41,73 @@ public ODataInnerError(Exception exception)
         {
             ExceptionUtils.CheckArgumentNotNull(exception, "exception");
 
-            this.Message = exception.Message ?? string.Empty;
-            this.TypeName = exception.GetType().FullName;
-            this.StackTrace = exception.StackTrace;
-
             if (exception.InnerException != null)
             {
                 this.InnerError = new ODataInnerError(exception.InnerException);
             }
+
+            Properties = new Dictionary<string, ODataValue>();
+
+            //// NOTE: we add empty elements if no information is provided for the message, error type and stack trace
+            ////       to stay compatible with Astoria.
+            Properties.Add(JsonConstants.ODataErrorInnerErrorMessageName, exception.Message.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorTypeNameName, exception.GetType().FullName.ToODataValue() ?? new ODataNullValue());
+            Properties.Add(JsonConstants.ODataErrorInnerErrorStackTraceName, exception.StackTrace.ToODataValue() ?? new ODataNullValue());
+        }
+
+        /// <summary>
+        ///  Initializes a new instance of the <see cref="T:Microsoft.OData.ODataInnerError" /> class with a dictionary of property names and corresponding ODataValues.
+        /// </summary>
+        /// <param name="properties">Dictionary of string keys with ODataValue as value. Key string indicates the property name where as the value of the property is encapsulated in ODataValue.</param>
+        /// <param name="nestedInnerErrorName">The name of the field for the nested inner error object inside inner error.</param>
+        /// <param name="nestedObject">Nested inner error object.</param>
+        public ODataInnerError(IDictionary<string, ODataValue> properties)

```

Comment:
>ODataInnerError(IDictionary<string, ODataValue> properties) [](start = 15, length = 59)

Do we have a scenario where we need this constructor?

Reply:
Yes, there is a scenario. With this constructor, stacktrace and other properties don't get auto added, therefore this will be the one getting used in the error handling lib. 

---
In reply to: [321856768](https://github.com/OData/odata.net/pull/1523#discussion_r321856768) [](ancestors = 321856768)

Reply:
See below. I don't think we need both this and the method that creates the empty ODataInnerError. Also, I'm not sure that the expected semantics would be that we would hold a reference to the dictionary, versus copy from it - does the user expect that we'll muck with their collection (i.e., if the user sets the Message property) and do they expect that if they continue to modify the dictionary that will be reflected in the inner error object?  Consider that, if we don't support ODataCollectionValue properties, we'll want to wrap the setter, which doesn't do us any good if they have direct access to the dictionary.

---
In reply to: [321868334](https://github.com/OData/odata.net/pull/1523#discussion_r321868334) [](ancestors = 321868334,321856768)

Reply:
I have fixed the part where top level dictionary does not get updated. However, if the dictionary contained references, the copy will reflect changes to those objects though. 

---
In reply to: [321906011](https://github.com/OData/odata.net/pull/1523#discussion_r321906011) [](ancestors = 321906011,321868334,321856768)